src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ArrayBuffer.cs

```diff
@@ -17,46 +18,50 @@ namespace System.Net.Http
     // Commit(byteCount) will extend the ActiveSpan by [byteCount] bytes into the AvailableSpan.
     // Discard(byteCount) will discard [byteCount] bytes as the beginning of the ActiveSpan.
 
-    // TODO: ISSUE 31300: Use ArrayPool to pool buffers.
-
+    [StructLayout(LayoutKind.Auto)]
     internal struct ArrayBuffer : IDisposable
     {
+        private readonly bool _usePool;
         private byte[] _bytes;
         private int _activeStart;
         private int _availableStart;
 
         // Invariants:
         // 0 <= _activeStart <= _availableStart <= bytes.Length
 
-        public ArrayBuffer(int initialSize)
+        public ArrayBuffer(int initialSize, bool usePool = false)
```

Comment:
What's the rationale for not always using the pool here?

Reply:
Two things (mentioned in the commit message):
- the pool is generally best for short usage. These buffers end up getting taken out of general circulation for a long time.
- when I tried disposing this when disposing the connection, it ended up triggering asserts on shutdown, with discards accessing the buffers even after disposal (maybe another issue lurking there)

Reply:
Yeah, Dispose on Http2Connection is weird because it doesn't actually completely dispose the connection. Rather, it initiates a graceful shutdown.

If I recall correctly, it was done this way to match how HttpConnection works. But arguably both aren't really correct and should not use Dispose for this purpose. 

We should probably clean this up, but it's not trivial to do.

Reply:
> * the pool is generally best for short usage

Is there any downside to using pooled buffers for a long time? Seems like there are advantages to pooling even when the buffer is in use for a long time.

Reply:
The right place to dispose the connection-level ArrayBuffers is here: https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1026

Reply:
> The right place to dispose the connection-level ArrayBuffers is here

That's where I did it.

Reply:
Hmmm, yeah, that's not the right place. 

The read buffer can be freed when ProcessIncomingFrames returns.

The write buffer is messier, I'm not sure where it should be freed exactly...

Reply:
I'm going to leave this as-is for now.